src/txn/txn.c

```diff
@@ -157,12 +156,7 @@ __wt_txn_get_snapshot(WT_SESSION_IMPL *session)
 
 	/* Walk the array of concurrent transactions. */
 	WT_ORDERED_READ(session_cnt, conn->session_cnt);
-	ckpt_id = txn_global->checkpoint_id;
 	for (i = n = 0, s = txn_global->states; i < session_cnt; i++, s++) {
-		/* Skip the checkpoint transaction; it is never read from. */
```

Comment:
That's right: instead of making every scan of `txn_global->states` check if each slot is doing a checkpoint, I've changed the checkpoint transaction to clear out its slot once it sets `txn_global->checkpoint_pinned` (and associated state).